node/consensus/src/lib.rs

```diff
@@ -65,7 +71,7 @@ pub struct Consensus<N: Network> {
     /// The unconfirmed solutions queue.
     solutions_queue: Arc<Mutex<IndexMap<PuzzleCommitment<N>, ProverSolution<N>>>>,
     /// The unconfirmed transactions queue.
-    transactions_queue: Arc<Mutex<IndexMap<N::TransactionID, Transaction<N>>>>,
+    transactions_queue: Arc<Mutex<BTreeMap<TransactionKey<N>, Transaction<N>>>>,
```

Comment:
This may not catch duplicate transactions that come in at different times. We may now have multiple instances of the same transaction in the `transactions_queue`.

Reply:
I see, good catch.

Just for shared context: Before adding to the queue, we already check if the tx was seen recently in an LRU cache: `self.seen_transactions.lock().put(transaction_id, ()).is_some()`. But `seen_transactions` can evict txs before they are popped from the queue, causing duplicate transactions to be added to the queue, passed to workers to verify, wasting compute.

Given that we hold on to *all* transactions in `transactions_queue` anyway, how about we make `seen_transactions` an `IndexSet` and we only evict once we add a transaction to the ledger. This requires more space but asymptotically is still O(N). 030dc1dbb implements this, and fixes a related bug where we didn't evict before reinsert_transmission.

This also points to another existing problem, namely that we didn't have any limit on the size of `transactions_queue`. Someone could spam e.g. 64_000 unique 1MB deployments and make the validators go OOM. Solutions don't have this issue because they are constant size and still have the LRU cache. I added a limit of ~1 GB: d1def3b54